doc/user_guide/display.rst

```diff
@@ -0,0 +1,307 @@
+frontend user-interface you are using.
+
+.. _renderers:
+
+Renderers
+---------
+
+Altair displays visualizations using renderers. There are two aspects of renderers:
+
+* The Python side :ref:`renderer-api` that sends display data to various frontend
+  user-interfaces.
+* Frontend user-interfaces that interpret that display data and render a visualiztion.
+  See relevant instructions for:
+
+  * :ref:`display-notebook`
+  * :ref:`display-jupyterlab`
+  * :ref:`display-nteract`
+
+.. _renderer-api:
+
+Renderer API
+~~~~~~~~~~~~
+
+In Altair, a renderer is any function that accepts a Vega-Lite or Vega
+visualization specification as a Python ``dict``, and returns a Python ``dict``
+in Jupyter's `MIME Bundle format
+<https://jupyter-client.readthedocs.io/en/stable/messaging.html#display-data>`_.
+The keys of the MIME bundle should be MIME types (such as ``image/png``) and the
+values should be the data for that MIME type (text, base64 encoded binary or
+JSON). The type signature of a renderer is thus::
+
+    def renderer(spec: dict) -> dict:
+        ...
+
+Altair and various user-interfaces have standardized on a couple of custom MIME types
+for Vega and Vega-Lite:
+
+* Vega-Lite 2.x: ``application/vnd.vegalite.v2+json``
+* Vega-Lite 1.x: ``application/vnd.vegalite.v1+json``
+* Vega 3.x: ``application/vnd.vega.v3+json``
+* Vega 2.x: ``application/vnd.vega.v2+json``
+
+The default renderers simply take a JSON spec and return a MIME bundle with one
+of these MIME types::
+
+    def default_renderer(spec):
+        bundle = {}
+        metadata = {}
+        bundle['text/plain'] = '<VegaLite object>`
+        bundle[mime_type] = 'application/vnd.vegalite.v2+json'
+        return bundle, metadata
+
+If a renderer needs to do custom display logic that doesn't use Jupyter's display
+system, it can return an empty MIME bundle ``dict``::
+
+    def non_jupyter_renderer(spec):
+        # Custom display logic that uses the spec
+        ...
+        # Return empty MIME bundle
+        return {}
+
+Altair offers an API to list the known renderers, register new ones and enable
+a given one. To return the registered renderers as a Python list::
+
+    >>> import altair as alt
+    >>> alt.renderers.names()
+    ['default', 'json']
+
+To enable the JSON renderer, which results in a collapsible JSON tree view
+in JupyterLab/nteract::
+
+    >>> alt.renderers.enable('json')
+
+To register and enable a new renderer::
+
+    >>> alt.renderers.register('custom_renderer', custom_renderer)
+    >>> alt.renderers.enable('custom_renderer')
+
+Renderers can also be registered using the `entrypoints`_ API of Python packages.
+For an example, see `ipyvega3`_.
+
+This same ``renderer`` objects exists separately on all of the Python APIs 
+for Vega-Lite/Vega described in :ref:`importing`.
+
+.. _display-notebook:
+
+Jupyter Notebook
+~~~~~~~~~~~~~~~~
+
+To render Vega-Lite 2.x or Vega 3.x in the Jupyter Notebook, you will need to
```

Comment:
We should somehow explicitly note here that most users will want Vega-Lite 2.X, because Altair users might not be aware of what the versions mean.